### PR TITLE
rgw/s3-notifications: async commit to persistent notification queue

### PR DESCRIPTION
**Pull request by igomon-bloomberg**
Notification reservations are now committed asynchronously (this only applies to persistent notifications).

Fixes: https://tracker.ceph.com/issues/63390
Signed-off-by: Igor Gomon igomon@bloomberg.net

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_notify.cc [hunk 1]
```diff
@@ -84,6 +84,26 @@ auto make_stack_allocator() {
...
+    PublishCommitCompleteArg(std::string _queue_name, const DoutPrefixProvider *_dpp)
+            : queue_name{std::move(_queue_name)}, dpp{_dpp} {}
+
+    std::string queue_name;
+    const DoutPrefixProvider *dpp;
+};
+
+void publish_commit_completion(rados_completion_t completion, void *arg) {
+    auto *comp_obj = static_cast<librados::AioCompletionImpl *>(completion);
+    auto *pcc_arg = static_cast<PublishCommitCompleteArg *>(arg);
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:98]: probably better to use an `std:unique_ptr` here.
also, can you dio static cast from `void*`? i think that reinterpret cast would work better here

File: src/rgw/driver/rados/rgw_notify.cc [hunk 2]
```diff
@@ -1087,14 +1106,18 @@ int publish_commit(rgw::sal::Object* obj,
       std::vector<buffer::list> bl_data_vec{std::move(bl)};
       librados::ObjectWriteOperation op;
       cls_2pc_queue_commit(op, bl_data_vec, topic.res_id);
-      const auto ret = rgw_rados_operate(
-	dpp, res.store->getRados()->get_notif_pool_ctx(),
-	queue_name, &op, res.yield);
+      auto *completion = librados::Rados::aio_create_completion();
+      auto *pcc_arg = new PublishCommitCompleteArg{queue_name, dpp};
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:1110]: probably safer to use use `std::unique_ptr` here as well. you would need to call `release()` if ` ret == 0` after line 1122 (to make sure we don't double delete) and pass the raw pointer to `set_complete_callback()`

File: src/rgw/driver/rados/rgw_notify.cc (hunk 1 above)
Comment (yuvalif) [reply src/rgw/driver/rados/rgw_notify.cc]: > I believe both can be used to convert from void* to T*

did not know that. thanks!

**Comment by yuvalif** [comment]
my main concern here is the case the the change to the queue is slower than the actual rados operations trigerring the notifications.
when `publish_commit()` was syncronouse it was just slowing down the operations, but with this change, the completion objects will just build up and eventually take all memeory.
can we add a counter for pending publish commit completions, and reply with "busy" error to the client if the counter crosses a threshold?

File: src/rgw/driver/rados/rgw_notify.cc (hunk 2 above)
Comment (igomon-bloomberg) [reply src/rgw/driver/rados/rgw_notify.cc:1110]: Right. I thought about that but somehow I thought we could have a race condition around the `pcc_arg` between `publish_commit()` and `publish_commit_completion()`. I see now that that's not an issue as the ret value from `aio_complete()` is a clear indicator of whether we should hold on to the pointer or `release()` it.  I will use unique_ptr for both pcc_arg and completion.

**Comment by yuvalif** [comment]
> my main concern here is the case the the change to the queue is slower than the actual rados operations trigerring the notifications. when `publish_commit()` was syncronouse it was just slowing down the operations, but with this change, the completion objects will just build up and eventually take all memeory. can we add a counter for pending publish commit completions, and reply with "busy" error to the client if the counter crosses a threshold?

had a discussion about that with @cbodley, the number of pending completions will not exceed the number of pending reservations. the `publish_reserve()` command is syncronouse with the opration and would fail the operation onc we ran out of space in the queue, even if this space is only reserved and not actually used.

**Comment by igomon-bloomberg** [comment]
I see. Do you have a suggestion of how we could define the threshold value?

**Comment by igomon-bloomberg** [comment]
> > my main concern here is the case the the change to the queue is slower than the actual rados operations trigerring the notifications. when `publish_commit()` was syncronouse it was just slowing down the operations, but with this change, the completion objects will just build up and eventually take all memeory. can we add a counter for pending publish commit completions, and reply with "busy" error to the client if the counter crosses a threshold?
> 
> had a discussion about that with @cbodley, the number of pending completions will not exceed the number of pending reservations. the `publish_reserve()` command is syncronouse with the opration and would fail the operation onc we ran out of space in the queue, even if this space is only reserved and not actually used.

Ok, I see. So we are then good to go without the threshold check?

**Comment by yuvalif** [comment]
> > > my main concern here is the case the the change to the queue is slower than the actual rados operations trigerring the notifications. when `publish_commit()` was syncronouse it was just slowing down the operations, but with this change, the completion objects will just build up and eventually take all memeory. can we add a counter for pending publish commit completions, and reply with "busy" error to the client if the counter crosses a threshold?
> > 
> > 
> > had a discussion about that with @cbodley, the number of pending completions will not exceed the number of pending reservations. the `publish_reserve()` command is syncronouse with the opration and would fail the operation onc we ran out of space in the queue, even if this space is only reserved and not actually used.
> 
> Ok, I see. So we are then good to go without the threshold check?

yes

**Comment by yuvalif** [comment]
* please squash commits
* please run it against the `rgw:notifications` teuthology suite

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
@yuvalif @mkogan1 this should be a nice performance boost. do you think it'd be worth trying to measure that?

**Comment by yuvalif** [comment]
> @yuvalif @mkogan1 this should be a nice performance boost. do you think it'd be worth trying to measure that?

Yes. This isnthe plan once the code is finalized

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
perf test instructions are here: https://gist.github.com/yuvalif/5587af49ebed90d2bed10d3312223e5c

> could not find a difference between a difference in the drop of performence between "main" and this branch. in both cases IO/s drop from ~1200 to ~750

**Comment by mkogan1** [comment]
> perf test instructions are here: https://gist.github.com/yuvalif/5587af49ebed90d2bed10d3312223e5c
> 
> > could not find a difference between a difference in the drop of performence between "main" and this branch. in both cases IO/s drop from ~1200 to ~750 for `PUT` and from ~1300 to ~850 for `DEL`

although there is no perceived difference in `hsbench` workload performance between this PR and main 
(it is probably capped by disk IO utilization)
few things that can think of - if we also compare the before and after this PR of
- latency (hsbench `Lat(ms): [ min: ..., avg: ..., 99%: ..., max: ...]` )
- radosgw process CPU utilization

one of those metrics could show an advantage of this commit  
(myb there are other notification related matrices worth evaluating)

**Comment by cbodley** [comment]
> perf test instructions are here: https://gist.github.com/yuvalif/5587af49ebed90d2bed10d3312223e5c
> 
> > could not find a difference between a difference in the drop of performence between "main" and this branch. in both cases IO/s drop from ~1200 to ~750 for `PUT` and from ~1300 to ~850 for `DEL`

thanks @yuvalif. i'm a bit confused by the TopicConfiguration there. doesn't `"Events": []` mean that it won't actually match any write/delete requests?

**Comment by yuvalif** [comment]
> > perf test instructions are here: https://gist.github.com/yuvalif/5587af49ebed90d2bed10d3312223e5c
> > > could not find a difference between a difference in the drop of performence between "main" and this branch. in both cases IO/s drop from ~1200 to ~750 for `PUT` and from ~1300 to ~850 for `DEL`
> 
> thanks @yuvalif. i'm a bit confused by the TopicConfiguration there. doesn't `"Events": []` mean that it won't actually match any write/delete requests?

yes, if no events are specified, we notify on all creation and deletion events

**Comment by cbodley** [comment]
> yes, if no events are specified, we notify on all creation and deletion events

i see. is that special behavior for rgw? i didn't see anything like that from aws docs

**Comment by yuvalif** [comment]
# CPU Utilization (baseline)
cpu-main
# hsbench Average Latency (ms) and  IO/s (baseline)
hsbench-main

**Comment by yuvalif** [comment]
# CPU Utilization (this PR):
cpu1
# hsbench Average Latency (ms) and IO/s (this PR):
hsbench-async

**Comment by igomon-bloomberg** [comment]
That is surprising. I did see clear reduction in latencies and increased PUT/s in my local testing. Overall, the penalty for enabling bucket notification was up to 40% less for object sizes up to 4K with this PR. I used warm put benchmark running against a VM cluster. We are also going to do performance testing against a cluster running on a bare metal machines soon.

**Comment by mkogan1** [comment]
@yuvalif, regarding why performance dosent scale above 2K IOPS , it is because 
```
hsbench -a 0555b35654ad1656d804 -s h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q== \
  -u http://localhost:8000/ -bp bk -m ipd -t 64 -z 1K -l 100 -ri 10 -d 120
 ```
is operation on a single bucket, lets please try to increase the number of buckets, ex add  `-b 6` to see if it scales up
(6*2000 = 12000 IOPS, this system has performed 11683 IOPS in my hsbench 4KB PUT test run)

**Comment by mkogan1** [comment]
updating for future reference, worked with @yuvalif on slack and we found that the PGs number had to be increased to get nominal performance:
`-o osd_pool_default_pg_num=128 -o osd_pool_default_pgp_num=128 -o mon_max_pg_per_osd=32768 -o mon_pg_warn_max_per_osd=32768 -o osd_pool_default_pg_autoscale_mode=warn`

**Comment by yuvalif** [comment]
> > yes, if no events are specified, we notify on all creation and deletion events
> 
> i see. is that special behavior for rgw? i didn't see anything like that from aws docs

@cbodley you are right, but this is the way we configure that in the past 4 years, so probably not going to change that...

**Comment by yuvalif** [comment]
> That is surprising. I did see clear reduction in latencies and increased PUT/s in my local testing. Overall, the penalty for enabling bucket notification was up to 40% less for object sizes up to 4K with this PR. I used warm put benchmark running against a VM cluster. We are also going to do performance testing against a cluster running on a bare metal machines soon. My measurement is based on the latency attribute of the PUT op response as reported in Ceph's logs.

@igomon-bloomberg  could you provide more details on how to reproduce your results?

**Comment by cbodley** [comment]
> @cbodley you are right, but this is the way we configure that in the past 4 years, so probably not going to change that...

i opened https://tracker.ceph.com/issues/64253 to track this. it doesn't look like we document that specific behavior, so i think we should treat it as a bug fix

**Comment by yuvalif** [comment]
> > @cbodley you are right, but this is the way we configure that in the past 4 years, so probably not going to change that...
> 
> i opened https://tracker.ceph.com/issues/64253 to track this. it doesn't look like we document that specific behavior, so i think we should treat it as a bug fix

actually, we have. in: https://docs.ceph.com/en/latest/radosgw/s3/bucketops/#id22
"
List of supported events see: [S3 Notification Compatibility](https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility). Multiple Event entities can be used. 
***If omitted, all “Created” and “Removed” events are handled.*** “
Lifecycle” and “Synced” event types must be specified explicitly.
"

**Comment by yuvalif** [comment]
running at ~7K IO/s did not seem to change the CPU patterns.
# Basline (without and then with notifications):
cpu-baseline
# this PR (without and then with notifications):
cpu

**Comment by igomon-bloomberg** [comment]
> > That is surprising. I did see clear reduction in latencies and increased PUT/s in my local testing. Overall, the penalty for enabling bucket notification was up to 40% less for object sizes up to 4K with this PR. I used warp put benchmark running against a VM cluster. We are also going to do performance testing against a cluster running on a bare metal machines soon. My measurement is based on the latency attribute of the PUT op response as reported in Ceph's logs.
> 
> @igomon-bloomberg could you provide more details on how to reproduce your results?

@yuvalif my test setup is a single-site VM cluster (2 RGWs, 3 MONs, 3 OSDs). The command I used was:
warp put --obj.size=<1K/4K> --access-key=.. --secret-key=... --host=.. --bucket=test-bucket
The notifications were sent to a Kafka endpoint.
I calculated average latency over all PUT requests as reported by latency attribute value in Ceph logs. 
The results I got for my single-site cluster are:

WITHOUT PR:
Obj Size |  Notifications On | Notifications Off
---------------------------------------------
256        |  54ms                   | 28ms
1K           |  56ms                   | 29ms
4K          |  57ms                    | 30ms

WITH PR:
Obj Size |  Notifications On | Notifications Off
---------------------------------------------
256        |  45ms                   | 32ms
1K           |  47ms                   | 29ms
4K          |  50ms                    | 31ms

However I did more tests recently (against a multisite cluster) and I noticed that latencies jump much higher when notifications are enabled. I only recorded numbers for 4K:
WITHOUT PR:
Obj Size |  Notifications On | Notifications Off
---------------------------------------------
4K       |  188ms                   | 32ms

WITH PR:
Obj Size |  Notifications On | Notifications Off
---------------------------------------------
4K       |  187ms                   | 29ms

Not quite what I was expecting and seems to me like maybe less CPU/disk capacity is available due to multisite processing?

**Comment by yuvalif** [comment]
would say that since the fix in the PR seeme correct logically, and it is quite clear that it does not create any performance regression, we should merge that (even if we could not see improvement in all tests) once we pass qa.

**Comment by cbodley** [comment]
passed qa after rerun https://pulpito.ceph.com/cbodley-2024-02-10_20:17:29-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**